### PR TITLE
Update network: faster training with AdamW optimizer

### PR DESCRIPTION
**Pull request by dubslow**
The resulting net is less sparse and thus a little slower than before, but the better eval prevails at VVLTC. The reduction in training time aids future net experiments.

The training occurred at https://github.com/vondele/nettest/pull/448
...using Tony's AdamW optimizer PR: https://github.com/official-stockfish/nnue-pytorch/pull/511

The main goal was to reduce resource usage during training.

As the currently used ranger21 variant uses lookahead and "stable weight decay", the training with adamw is around 5% faster (no stable weight decay) and uses less memory during training and for checkpoints (no lookahead).

The other effects came as somewhat of a surprise.

It is unclear what exactly causes the lower sparsity, around 50-52% vs 60+% which does cause a significant slowdown. It is also unclear what exactly causes the apparent better scaling.

Failed STC nonreg: https://tests.stockfishchess.org/tests/view/6a6e1580c054e285ae029c2e
LLR: -2.94 (-2.94,2.94) <-1.75,0.25>
Total: 376800 W: 97403 L: 98082 D: 181315
Ptnml(0-2): 1226, 44985, 96614, 44392, 1183

Nodestime STC nonreg: https://tests.stockfishchess.org/tests/view/6a72310d2cf557d58a2e1ced
LLR: 2.96 (-2.94,2.94) <-1.75,0.25>
Total: 22812 W: 6130 L: 5891 D: 10791
Ptnml(0-2): 89, 2537, 5920, 2766, 94

LTC: https://tests.stockfishchess.org/tests/view/6a7205ef2cf557d58a2e1cb8
Ptnml(0-2): 135, 31150, 81387, 31468, 145
LLR -0.64 <0.00,2.00>
LLR: 7.72 <-1.75,0.25>

Passed VLTC: https://tests.stockfishchess.org/tests/view/6a7a7b595ebead9b2e786ee7
LLR: 2.95 (-2.94,2.94) <0.00,2.00>
Total: 139242 W: 36650 L: 36248 D: 66344
Ptnml(0-2): 13, 13245, 42711, 13631, 21

Passed VVLTC: https://tests.stockfishchess.org/tests/view/6a72cbb02cf557d58a2e1e40
LLR: 2.95 (-2.94,2.94) <0.50,2.50>
Total: 71720 W: 18989 L: 18641 D: 34090
Ptnml(0-2): 6, 6229, 23046, 6569, 10

bench 2522345

### CODE REVIEW

**Comment by dubslow** [comment]
I think the 3 mutually exclusive scaling tests (red -> blue -> strong green) make a pretty persuasive case for merging on generally gainer grounds. (The fourth nodestime test is icing on the cake.)

That said, per Discord, there is some doubt on the matter, or if it counts as a non-reg sort of PR or whatever, so I've launched a [second test intended to pass green](https://tests.stockfishchess.org/tests/view/6a7a7b595ebead9b2e786ee7).

**Comment by TonyCongqianWang** [comment]
The main goal was to reduce ressource usage during training.

As the currently used ranger21 variant uses lookahead and "stable weight decay", the training with adamw is around 5% faster (no stable weight decay) and uses less memory during training and for checkpoints (no lookahead).

The other effects came at somewhat of a surprise.

It is unclear what exactly causes the lower sparsity. The sparsity is now around 50-52% vs 60+% which does cause a significant slowdown.

It is also unclear what exactly causes the apparent better scaling.

**Comment by mstembera** [comment]
Just out of curiosity regarding the lower sparsity...  Do we use a type of regularizer in the trainer to promote sparsity?

**Comment by TonyCongqianWang** [comment]
> Just out of curiosity regarding the lower sparsity...  Do we use a type of regularizer in the trainer to promote sparsity?

Good question. We do not use any explicit loss to incentive sparsity... The optimiser simply did it. Though learning rate seems to affect it -- given that everything else stays the same

**Comment by dubslow** [comment]
The second gainer test passed, what a surprise.  Now a clear progression of red -> blue -> weak green -> strong green.

I've force-pushed with a commit message reflecting Tony's comments.

**Comment by vondele** [comment]
> I've force-pushed with a commit message reflecting Tony's comments.

you'll need to edit the initial PR message to reflect that, since we now merge based on that.

**Comment by dubslow** [comment]
> you'll need to edit the initial PR message to reflect that, since we now merge based on that.

Oh wow that's news to me. Github does squash-merges now...?

**Comment by Disservin** [comment]
> > you'll need to edit the initial PR message to reflect that, since we now merge based on that.
> 
> Oh wow that's news to me. Github does squash-merges now...?

our merge scripts take the commit message from the PR body, since that's where people write the info most of the time

**Comment by dubslow** [comment]
> our merge scripts take the commit message from the PR body, since that's where people write the info most of the time

the art of git replaced with the art of the web browser

/cloudyelling

(it does make it easier to edit the message without retriggering CI a bunch)

**Comment by vondele** [comment]
> the art of git replaced with the art of the web browser

the art of the perfect pull request.. one commit, no force pushes. In this case, both are the same automatically.

**Comment by TonyCongqianWang** [comment]
Huh? When did this change. Wasnt manual squashing something I got bothered to do in one of my PRs lol
